File: api/data_loader.py
# api/data_loader.py

import pandas as pd
import pickle
import os
from api.logger import logger  # Import the logger

def load_dataset(file_path: str, cache_file_path: str) -> pd.DataFrame:
    """
    Loads the TEDx dataset with caching.

    Args:
        file_path (str): Path to the CSV file.
        cache_file_path (str): Path to the cache file.

    Returns:
        pd.DataFrame: Loaded dataset.
    """
    logger.info("Loading the TEDx Dataset with a caching mechanism.")
    data = pd.DataFrame()

    if os.path.exists(cache_file_path):
        logger.info("Loading cached dataset.")
        try:
            with open(cache_file_path, 'rb') as cache_file:
                data = pickle.load(cache_file)
            logger.info(f"Cached dataset loaded successfully with {len(data)} records.")
        except Exception as e:
            logger.error(f"Error loading cached dataset: {e}")
    else:
        logger.info("Loading dataset from CSV file.")
        try:
            data = pd.read_csv(file_path)
            logger.info(f"Dataset successfully loaded with {len(data)} records.")
            # Cache the dataset for future use
            with open(cache_file_path, 'wb') as cache_file:
                pickle.dump(data, cache_file)
            logger.info("Dataset cached successfully.")
        except Exception as e:
            logger.error(f"Error loading dataset: {e}")

    return data

File: app/api/search/data/dataLoader.ts
// File: app/api/search/data/dataLoader.ts

import { loadCache, saveCache } from '../cache/cacheManager';
import { parse } from 'csv-parse/sync';
import fs from 'fs/promises';

type TEDxTalk = {
  slug: string;
  description: string;
  presenterDisplayName: string;
  sdg_tags?: string[];
  description_vector?: number[];
};

export async function loadData(): Promise<TEDxTalk[]> {
  // Placeholder for your actual data loading logic
  const data: TEDxTalk[] = []; // Replace this with your actual data loading logic
  return data;
}

export async function loadDataset(filePath: string, cacheFilePath: string): Promise<TEDxTalk[]> {
  console.log('Loading the TEDx dataset with caching.');
  let data: TEDxTalk[] | null = await loadCache<TEDxTalk[]>(cacheFilePath);

  if (data) {
    console.log('Dataset loaded from cache.');
  } else {
    console.log('Loading dataset from CSV file.');
    const fileContent = await fs.readFile(filePath, 'utf-8');
    data = parse(fileContent, { columns: true }) as TEDxTalk[];
    
    // Save parsed data to cache
    await saveCache(data, cacheFilePath);
    console.log(`Dataset parsed and cached at ${cacheFilePath}.`);
  }
  return data;
}
